File: AC-tually Cold Enough/d1.java
import java.io.*;
import java.util.*;

public class d1 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] first = br.readLine().split(" ");
        
        int N = Integer.parseInt(first[0]);
        int D = Integer.parseInt(first[1]);
        int K = Integer.parseInt(first[2]);
        
        long[] H = new long[N];
        String[] hVals = br.readLine().split(" ");
        for (int i = 0; i < N; i++) {
            H[i] = Long.parseLong(hVals[i]);
        }
        
        ArrayList<Integer>[] adj = new ArrayList[N];
        for (int i = 0; i < N; i++) {
            adj[i] = new ArrayList<>();
        }
        
        for (int i = 0; i < N - 1; i++) {
            String[] edge = br.readLine().split(" ");
            int u = Integer.parseInt(edge[0]);
            int v = Integer.parseInt(edge[1]);
            adj[u].add(v);
            adj[v].add(u);
        }
        
        // Build parent-child relationships
        int[] parent = new int[N];
        Arrays.fill(parent, -1);
        ArrayList<Integer>[] children = new ArrayList[N];
        for (int i = 0; i < N; i++) {
            children[i] = new ArrayList<>();
        }
        
        Queue<Integer> q = new LinkedList<>();
        q.add(0);
        parent[0] = 0;
        
        while (!q.isEmpty()) {
            int u = q.poll();
            for (int v : adj[u]) {
                if (parent[v] == -1) {
                    parent[v] = u;
                    children[u].add(v);
                    q.add(v);
                }
            }
        }
        
        if (K > N) {
            System.out.println(-1);
            return;
        }
        
        // Check if possible at all
        long maxH = 0;
        for (long h : H) {
            maxH = Math.max(maxH, h);
        }
        long high = maxH + D * N + 5;
        
        // Binary search
        long low = 0;
        long ans = -1;
        
        // Check with max possible T first
        if (!canCool(high, N, D, K, H, children)) {
            System.out.println(-1);
            return;
        }
        
        while (low <= high) {
            long mid = low + (high - low) / 2;
            
            if (canCool(mid, N, D, K, H, children)) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        
        System.out.println(ans);
    }
    
    static boolean canCool(long T, int N, int D, int K, long[] H, ArrayList<Integer>[] children) {
        int cooled = 0;
        
        // Stack for DFS: pairs of (node, temperature)
        Stack<Pair> stack = new Stack<>();
        stack.push(new Pair(0, T));
        
        while (!stack.isEmpty()) {
            Pair p = stack.pop();
            int u = p.node;
            long temp = p.temp;
            
            if (temp >= H[u]) {
                cooled++;
                if (cooled >= K) return true;
                
                for (int v : children[u]) {
                    stack.push(new Pair(v, temp - D));
                }
            }
            // If not cooled, don't propagate
        }
        
        return cooled >= K;
    }
    
    static class Pair {
        int node;
        long temp;
        Pair(int node, long temp) {
            this.node = node;
            this.temp = temp;
        }
    }
}

File: AC-tually Cold Enough/d1.py
import sys
from collections import deque

input = sys.stdin.readline

N, D, K = map(int, input().split())
H = list(map(int, input().split()))

adj = [[] for _ in range(N)]
for _ in range(N - 1):
    u, v = map(int, input().split())
    adj[u].append(v)
    adj[v].append(u)

# Build tree structure with parent-child relationships
parent = [-1] * N
children = [[] for _ in range(N)]

q = deque([0])
parent[0] = 0

while q:
    u = q.popleft()
    for v in adj[u]:
        if parent[v] == -1:
            parent[v] = u
            children[u].append(v)
            q.append(v)

if K > N:
    print(-1)
    sys.exit()

def can_cool(T):
    cooled = 0
    stack = [(0, T)]  # (node, temperature)
    
    while stack:
        u, temp = stack.pop()
        
        if temp >= H[u]:
            cooled += 1
            if cooled >= K:
                return True
            
            # Only propagate to children if this room is cooled
            for v in children[u]:
                stack.append((v, temp - D))
    
    return cooled >= K

# Check if possible at all
maxH = max(H)
high = maxH + D * N + 5

if not can_cool(high):
    print(-1)
    sys.exit()

# Binary search for minimum T
low = 0
ans = -1

while low <= high:
    mid = (low + high) // 2
    
    if can_cool(mid):
        ans = mid
        high = mid - 1
    else:
        low = mid + 1

print(ans)